### PR TITLE
Add API documentation for 'PreferenceService'

### LINKED ISSUE #8611: Add Documentation for 'PreferenceService' and 'PreferenceProxy'
Add API documentation to `PreferenceService` and related concepts (like `PreferenceProxy`). This will help developers to better understand how to interact with and use `PreferenceService` as well as `createPreferenceProxy`.

### PR DESCRIPTION
**Pull request by sdirix**
#### Summary
Document 'PreferenceService', 'PreferenceProxy' and related interfaces,
classes and functions.

Closes #8611

Signed-off-by: Stefan Dirix <sdirix@eclipsesource.com>
Contributed on behalf of STMicroelectronics

#### What it does
 * Add API documentation to 'PreferenceService', 'PreferenceProxy' and more
 * Minor backwards compatible typing adjustments in `preference-service.ts` for cleanup and documentation purposes.

#### How to test
 * Check documentation and make sure everything compiles

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

File: packages/core/src/browser/preferences/preference-proxy.ts [hunk 1]
```diff
@@ -33,21 +33,103 @@ export interface PreferenceEventEmitter<T> {
...
+ *
+ * type MyPreferences = PreferenceProxy<MyPrefConfiguration>;
+ *
+ * const preferences : MyPreferences = createPreferenceProxy(preferenceService, MyPreferenceSchema);
+ *
+ * preferences.onPreferenceChanged(({ preferenceName, newValue }) => { console.log(preferenceName, newValue) });
+ * const enabled = preferences['myext.decorations.enabled'];
+ * ```
+ *
+ */
```
Comment (vince-fugnitto) [inline packages/core/src/browser/preferences/preference-proxy.ts:93]: Would it be better to link to an actual example in the codebase (ex: [core-preferences.ts](https://github.com/eclipse-theia/theia/blob/master/packages/core/src/browser/core-preferences.ts))?
My reasoning for this approach would be that:
- the actual example would be maintained and updated, while the example provided might get outdated.
- the actual example is syntax highlighted (helps the readability).
- the actual example describes how to bind it to the container which is not covered here.

I'm of course open to suggestions on whether we should keep it this way or not.

File: packages/core/src/browser/preferences/preference-service.ts [hunk 2]
```diff
@@ -66,34 +66,163 @@ export interface PreferenceChanges {
 }
 
 export const PreferenceService = Symbol('PreferenceService');
+/**
+ * Service to store and retrieve preference values.
```
Comment (vince-fugnitto) [inline packages/core/src/browser/preferences/preference-service.ts:70]: The description here might not be accurate, while it does allow clients to store and retrieve preference values, they are also able to listen to preference change events in addition to other utilities.

File: packages/core/src/browser/preferences/preference-proxy.ts [hunk 3]
```diff
@@ -33,21 +33,103 @@ export interface PreferenceEventEmitter<T> {
...
+ * ```ts
+ * interface PreferenceConfiguration {
+ *  'myext.decorations.enabled': boolean,
+ *  'myext.debug': boolean,
+ *  'myext.editor.previewtype': string
+ * }
+ * const prefs : PreferenceRetrieval<PreferenceConfiguration> = createPreferenceProxy(...);
+ * const enabled : boolean = prefs.get('myext.decorations.enabled'); // valid
+ * const debug : string = prefs.get('myext.debug'); // invalid
+ * prefs.get('foobar'); // invalid
```
[the remark this answers is not in the document]
Comment (sdirix) [reply packages/core/src/browser/preferences/preference-proxy.ts:51]: I added this example mainly to showcase what the `PreferenceRetrieval<T>` is used for. It's true that in practice a developer using the preference proxy will not use the interface that way. I'll generalize it a bit.

File: packages/core/src/browser/preferences/preference-proxy.ts (hunk 1 above)
Comment (sdirix) [reply packages/core/src/browser/preferences/preference-proxy.ts:93]: I very much like the idea to link to an actual example. However I still think that this documentation here has some merit as it distills the usage down to what really matters in regards to the `PreferenceProxy`. I tried to not give a whole set of instructions of what the user can do but what they really need in practice, i.e. how to create and access the preferences. For me the binding was out of scope but I see how one can claim that it's missing.

Regarding the syntax highlighting: When hovering over the `PreferenceProxy` type the documentation shown in the popup is actually highlighted.

I'll reword it a bit and link to a real example.

File: packages/core/src/browser/preferences/preference-service.ts (hunk 2 above)
Comment (sdirix) [reply packages/core/src/browser/preferences/preference-service.ts:70]: That's true, the description is too simplistic.

File: packages/core/src/browser/preferences/preference-proxy.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (sdirix) [reply packages/core/src/browser/preferences/preference-proxy.ts]: I removed the `createPreferenceProxy` call. I think this was the main misleading line and the intention is now clearer.

File: packages/core/src/browser/preferences/preference-proxy.ts (hunk 1 above)
Comment (sdirix) [reply packages/core/src/browser/preferences/preference-proxy.ts]: I removed the code example and replaced it with the general principle behind it while linking to the `CorePreferences` (also verified that the link works in the generated docs).

File: packages/git/src/browser/git-preferences.ts [hunk 4]
```diff
@@ -66,7 +66,9 @@ export function createGitPreferences(preferences: PreferenceService): GitPrefere
 export function bindGitPreferences(bind: interfaces.Bind): void {
     bind(GitPreferences).toDynamicValue(ctx => {
         const preferences = ctx.container.get<PreferenceService>(PreferenceService);
-        return createGitPreferences(preferences);
+        const prefs = createGitPreferences(preferences);
+        prefs["git.editor.dirtyDiff.linesLimit"] = '20';
+        return prefs;
```
Comment (vince-fugnitto) [inline packages/git/src/browser/git-preferences.ts:71]: @sdirix are these changes related, they are currently failing the build.

Comment (sdirix) [reply packages/git/src/browser/git-preferences.ts:71]: I committed this file by mistake. I'll remove it again.

**Review by vince-fugnitto — APPROVED** [review]
The changes look very good! 
Thank you for the contribution @sdirix 👍

**Pull request merged**